borgmatic/commands/completion.py

```diff
@@ -55,3 +65,153 @@ def bash_completion():
             '\ncomplete -o bashdefault -o default -F complete_borgmatic borgmatic',
         )
     )
+
+
+# fish section
+
+
+def has_file_options(action: Action):
+    return action.metavar in (
+        'FILENAME',
+        'PATH',
+    ) or action.dest in ('config_paths',)
+
+
+def has_choice_options(action: Action):
+    return action.choices is not None
+
+
+def has_required_param_options(action: Action):
+    return (
+        action.required is True
+        or action.nargs
+        in (
+            '+',
+            '*',
+        )
+        or '--archive' in action.option_strings
+        or action.metavar in ('PATTERN', 'KEYS', 'N')
+        or (action.type is not None and action.default is None)
+    )
+
+
+def has_exact_options(action: Action):
+    return (
+        has_file_options(action) or has_choice_options(action) or has_required_param_options(action)
+    )
+
+
+def exact_options_completion(action: Action):
+    '''
+    Given an argparse.Action instance, return a completion invocation
+    that forces file completion or options completion, if the action
+    takes such an argument and was the last action on the command line.
+
+    Otherwise, return an empty string.
+    '''
+
+    if not has_exact_options(action):
+        return ''
+
+    args = ' '.join(action.option_strings)
+
+    if has_file_options(action):
+        return f'''\ncomplete -c borgmatic -Fr -n "__borgmatic_last_arg {args}"'''
+
+    if has_choice_options(action):
+        return f'''\ncomplete -c borgmatic -f -a '{' '.join(map(str, action.choices))}' -n "__borgmatic_last_arg {args}"'''
+
+    if has_required_param_options(action):
+        return f'''\ncomplete -c borgmatic -x -n "__borgmatic_last_arg {args}"'''
+
+    raise RuntimeError(
+        f'Unexpected action: {action} passes has_exact_options but has no choices produced'
```

Comment:
Given that this error message could make its way to a user, it might be a good idea to make it somewhat more user-centric (at the cost of specificity). E.g.: `Could not produce a tab completion for action {action}` or whatever. You can always include more information for the hapless debugging dev with a comment.

Reply:
can this error make it to a user? The unit test should fail and as far as I am aware actions are not created on the fly, I think this serves more as an impossible case where you forget to handle one of the cases you filtered for.

Reply:
Okay, got it.